### DISCUSSION TITLE
jakarta.validation, org.springframework.validation 차이

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#237] [Spring Core]
.

**Comment by gyungchan-jo** [comment]
### jakarta.validation

이 패키지는 자바 표준 유효성 검사 스펙(Jakarta Bean Validation)에 따른 API를 제공합니다. 이 스펙은 JSR-380 (Java Specification Request 380)으로 정의되며, 자바 기반 애플리케이션에서 데이터 모델의 유효성을 검사하는 데 사용됩니다.

주로 어노테이션을 사용하여 데이터 모델의 필드에 제약 조건을 정의하고, javax.validation.Validator 인터페이스를 통해 유효성 검사를 수행합니다.

jakarta.validation은 웹 애플리케이션뿐만 아니라 자바 애플리케이션 전반에 사용될 수 있으며, 여러 자바 프레임워크와 라이브러리에서 지원합니다.

### Spring Validation

이 패키지는 스프링 프레임워크 내에서 제공하는 유효성 검사 및 데이터 바인딩 관련 기능을 포함합니다. 스프링 프레임워크는 자바 애플리케이션의 다양한 측면을 다루며, 데이터 검증 및 바인딩 역시 그 중 하나입니다.

스프링은 org.springframework.validation.Validator 인터페이스와 org.springframework.validation.Errors 인터페이스를 사용하여 커스텀 유효성 검사를 정의하고 데이터 바인딩 오류를 처리합니다.

org.springframework.validation은 스프링 프레임워크에서 웹 애플리케이션과 관련된 유효성 검사 및 데이터 바인딩에 사용됩니다.

[참고 블로그](https://velog.io/@mercurios0603/jakarta.validation%EA%B3%BC-org.springframework.validation%EC%9D%98-%EC%B0%A8%EC%9D%B4)

**Comment by nayonsoso** [comment]
## jakarta.validation (Jakarta Bean Validation)
- 자바에서 제공하는 Specification이다. [공식문서](https://beanvalidation.org)
![image](https://github.com/user-attachments/assets/a6a2ac88-d5fb-4da7-b9f5-19f122676df6)

---

- Jakarta Bean Validation을 Specification을 구현한 모듈이 [Hibernate Validator](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#validator-gettingstarted-createproject)이다.
- Jakarta Bean Validation 는 명세이기 때문에 따로 구현체를 연결해서 사용하고 있고, Hibernate Validator 를 기본으로 사용하고 있다.


> Hibernate Validator is the reference implementation of Jakarta Bean Validation.

---

- 여러 레이어에서 반복되는 모델 검증로직의 중복을 줄이기 위해 등장했다.
- 단순히 프레젠트 레이어에서만 사용되는 목적은 아니다.

> It is specifically not tied to either web or persistence tier, and is available for both server-side application programming

![](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/images/application-layers.png)

![](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/images/application-layers2.png)

---

- 의존성을 추가해야 한다.

``` gradle
implementation('org.hibernate.validator:hibernate-validator:6.1.2.Final')  
implementation('org.glassfish:jakarta.el:3.0.3') // Java SE 환경이라면 추가
implementation 'org.springframework.boot:spring-boot-starter-validation' // 이걸 추가해도, 내부에 org.hibernate.validator를 포함하고 있으므로, 동작한다.
```

- cf. 만약 `implementation 'jakarta.validation:jakarta.validation-api:3.1.0'` 만 한다면, 아래와 같은 에러가 뜬다.
```
Unable to create a Configuration, because no Jakarta Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.
```

---

- 어노테이션을 사용해 필드의 제약조건을 설정한다.
  - 비어있지 않아야 함 : `@NotBlank`
  - 문자열의 크기가 최소한 몇 이상이여야 함 :`@Size(min = ?)`
- ExecutableValidator를 통해 검증한다. 
  - 어노테이션만 달고 끝나면 검증되지 않는다.

```java
class TestHaha {

    class Hi {
        @NotNull
        private final String message;

        public Hi(String message) {
            this.message = message;
        }
    }

    @Test
    void 그냥_어노테이션만_있으면_생성할_때_예외가_발생하지_않는다() {
        assertThatCode(() -> new Hi(null))
                .doesNotThrowAnyException();
    }

    @Test
    void 어노테이션_선언과_함께_validator를_사용하면_생성할_때_예외가_발생한다() {
        Hi hi = new Hi(null);
        ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory();
        Validator validator = validatorFactory.getValidator();
        Set<ConstraintViolation<Hi>> violations = validator.validate(hi);

        Assertions.assertThat(violations).isNotEmpty();
        Assertions.assertThatThrownBy(() -> {throw new ConstraintViolationException(violations);})
                .isInstanceOf(ConstraintViolationException.class);
    }
}
```

### org.springframework.validation
- 스프링이 제공한다.
- 커스텀 검증 로직으로 보다 복잡한 검증 조건을 구현할 수 있다.
- 구성 요소로는 Validator, Errors, BindingResult 등이 있다.
  - Validator : 커스텀 검증 로직을 구현하는데 사용된다.
  - Errors : 검증 도중 발생하는 오류들을 추적한다.
  - BindingResult : Errors의 구현체. 오류 정보를 담아 컨트롤러 메소드로 전달한다.

### 사용 예시
- 컨트롤러에서 jakarta.validation의 `@Valid` 어노테이션으로 ReqeustBody에 대한 검사를 할 수 있다.
- 그리고 검증 과정에서 예외가 발생하면 org.springframework.validation의 `BindingResult` 객체로 오류를 수집할 수 있다.


```java
@PostMapping("/users")
@ResponseBody
public void createUser(@Valid @RequestBody User user, BindingResult result) {
    if (result.hasErrors()) {
         // RequestBody 바인딩 시 발생한 예외를 처리할 수 있다.
    }
}
```


ref.

**Comment by 3Juhwan** [comment]
### jakarta.validation의 `@Valid`와 springframework.validation의 `@Validated`에 대하여

jakarta.validation은 `@Valid`와 `@Min`, `@Max` 같은 자바빈을 검증하는 어노테이션을 지원합니다. 일반적으로 DTO 필드에 검증 관련 어노테이션을 선언하고, 컨트롤러의 메서드 파라미터에 `@RequestBody`나 `@ModelAttribute`와 함께 `@Valid`를 선언합니다. 

위 예시에서 `@Valid`는 ArgumentResolver에 의해 처리됩니다. ArgumentResolver는 `@Valid`가 선언되어 있으면 유효성 검사를 진행합니다. 검증에 오류가 있다면 MethodArgumentNotValidException가 발생하고 DispatcherServlet에 등록된 기본 예외 리졸버인 DefaultHandlerExceptionResolver에 의해 400으로 처리됩니다. 이러한 이유로 `@Valid`는 컨트롤러에서만 동작합니다. 

다른 계층에서도 유효성 검사를 지원하기 위해 스프링은 AOP 기반의 유효성 검사를 지원합니다. 예를 들어, 서비스 클래스에 `@Validated`를 선언하고 검증할 메서드의 파라미터에 `@Valid`를 선언합니다. 그럼 스프링은 메서드 요청을 가로채 유효성 검증을 진행합니다. 유효성 검증에 실패하면 ConstraintViolationException를 발생하는데 ArgumentResolver에 의해 발생한 예외와 다른 것을 확인할 수 있습니다. 

[참고 자료](https://mangkyu.tistory.com/174)
